cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -194,7 +233,18 @@ public void dismissTransactionContext() {
 
   @Override
   public Iterable<TransactionAware> getTransactionAwares() {
-    return txContext == null ? NO_TX_AWARES : Iterables.concat(extraTxAwares, inProgress.values());
+    if (txContext == null) {
+      return NO_TX_AWARES;
+    }
+    ArrayList<TransactionAware> result = new ArrayList<>(extraTxAwares.size() + activeTxAwares.size());
+    result.addAll(extraTxAwares);
+    for (WeakReferenceDelegatingTransactionAware reference : activeTxAwares.values()) {
+      TransactionAware txAware = reference.get();
+      if (txAware != null) {
+        result.add(txAware);
```

Comment:
I am wonder, shall we copy the strong reference to a list and return or shall we return an transform+filter `Iterable`? The main difference is if we return the `Iterable`, it won't promote every still reachable references to strong reference right away (the TrasactionContext should hold a strong reference to those who actually participating in the transaction). The filter and transaction function will be invoked every time when `Iterable.iterator().next()` is called.

```java
Iterables.filter(
  Iterables.transform(activeTxAwares.values(), new Function<WeakReferenceDelegatingTransactionAware, TransactionAware>() {
  @Nullable
    public TransactionAware apply(WeakReferenceDelegatingTransactionAware ref) {
      return ref.get();
    }
  },
  Predicates.notNull());
```